### ISSUE TITLE
[RFC] Text Embedding Precomputation & Decoupled Loading for Minimax-H3

### ISSUE THREAD

**Issue by BeatSeat** [Tencent-Hunyuan/UniRL#450]
# [RFC] Text Embedding Precomputation & Decoupled Loading for Minimax-H3

## Motivation

In Flow-GRPO and related RL fine-tuning workflows for diffusion models, the text encoder displays extreme asymmetry in compute and lifecycle:
1. **Invoked only once during Rollout**: In `MiniMaxH3Pipeline.generate()`, prompts are encoded once via `self.text_embed.embed(texts)` to produce a `TextEmbedCondition`.
2. **Completely untouched during Replay and Backward**: `MiniMaxH3DiffusionStage.replay` directly reads `Part.conditions` attached to the sample. The text encoder sits completely idle during the entire training forward/backward passes.

However, for **Minimax-H3**, the text encoder is **Qwen3-VL-32B** (extracting layer-50 hidden states), which weighs **~64 GB** in `bfloat16`:
* **Host RAM explosion**: If kept on CPU, an 8-GPU worker node loads 8 separate instances, instantly exhausting **512 GB of host physical memory**.
* **GPU swapping & serialization bottlenecks**: If swapped dynamically onto GPU during Rollout, it triggers an exclusive node-level file lock (`_embedding_residency` using `fcntl.flock`), forcing all 8 ranks to serialize H2D weight transfers and VAE offloads, introducing severe multi-minute pipeline bubbles.
* **Ephemeral in-memory cache**: While `MiniMaxH3TextEmbedStage` implements an in-memory `_cache` of size 64, it is lost across runs and does not prevent the initial 512 GB memory allocation.

Since training/eval prompts in RL datasets (e.g. PickScore 10k, VBench) are known and static upfront, **precomputing text embeddings offline and bypassing the text encoder at training runtime** will:
* Save **512 GB of host RAM** (or 64 GB VRAM) per 8-GPU node.
* Eliminate all cross-rank `fcntl.flock` contention and sync barriers.
* Make all training ranks 100% symmetric and lean (`text_encoder = None`).

---

### Goals
* **Variable-Length Embeddings**: Store native, unpadded variable-length tensors to avoid wasting 70%+ disk space, batch-padded dynamically during collation.
* **Pipeline Bypass Protocol**: Enable `MiniMaxH3Pipeline` to consume pre-extracted `TextEmbedCondition` directly without invoking the 32B model.
* **Lean Bundle Loading**: Support `skip_text_encoder_loading: true` in `MiniMaxH3Bundle` to completely skip loading the 64 GB Qwen3-VL weights.
* **Robust Rerun & Invalidation Semantics**: Clearly define resume, append, force-overwrite, and fail-fast startup checks.

### Non-Goals
* **Architecture changes to DiT**: No modifications to diffusion transformer weights or attention structures.
* **Online runtime multi-process cache writing**: We avoid having multiple training ranks concurrently write cache files to prevent distributed filesystem lock contention, NFS `close-to-open` stale reads, and MDS metadata overload.

---

### 1. End-to-End Workflow

```
[Precomputation Stage]
    prompts (JSONL / Parquet / TXT)
               │
               ▼
    Precomputation Tool (Batch forward pass)
               │
               ▼
    Precomputed Artifacts (Variable-length tensors with metadata fingerprint)

===================================================================================

[Training Runtime]
    Recipe: use_precomputed_text_embeds: true
               │
               ├──> MiniMaxH3Bundle: Completely skips loading Qwen3-VL 32B
               │    (text_encoder = None -> Saves 512GB host RAM on 8-GPU node)
               │
               ├──> Dataset / DataLoader: Ingests prompt embedding
               │    (Applies dynamic padding per-batch to satisfy packed sequence geometry)
               │    (Injects TextEmbedCondition into sample.conditions)
               │
               └──> MiniMaxH3Pipeline.generate(): Bypass Protocol
                    (Detects TextEmbedCondition -> Bypasses text_embed.embed() -> 0ms latency)
```

---

### 2. Variable-Length Tensor Storage & Dynamic Collation

Minimax-H3's packed sequence requires identical token lengths within a single micro-batch (`len(lengths) == 1`). 

* **Why Variable-Length Storage?**
  * Padding every prompt to `max_seq_len = 512` offline wastes ~70% of storage and memory bandwidth, since typical prompts are only 40–120 tokens.
  * Therefore, offline precomputation will store **raw, unpadded 2D tensors** `[L_i, 5120]` alongside their sequence lengths `L_i`.
* **Dynamic Collation**:
  * The dataset collator groups prompts of equal lengths or dynamically pads each batch to the batch maximum `max(L_batch)` right before creating `TextEmbedCondition`.

---

### 3. Rerun & Lifecycle Mechanics ("What happens when re-running?")

To ensure experiment reproducibility and prevent silent errors:

#### A. Key Identity & Metadata Fingerprint
* **Sample Key**: `prompt_key = sha256(prompt.strip())[:16]`
* **Fingerprint Header**:
  ```json
  {
    "schema_version": "1.0",
    "extractor": "minimax_h3_text_embed",
    "model_checkpoint": "Qwen/Qwen2.5-VL-32B-Instruct",
    "model_hash": "<commit/weight_sha256>",
    "target_layer": 50,
    "feature_dim": 5120,
    "dtype": "bfloat16"
  }
  ```

#### B. Rerun & Update Policies
1. **Interrupted Runs (`--resume`)**: The extraction tool checks existing keys in the output index. Existing keys are skipped with 0 duplicate compute; only missing prompts are forwarded and appended.
2. **Dataset Expansions (Incremental Append)**: When prompts are added, the tool computes new keys and appends them without re-extracting existing ones.
3. **Weight/Layer Invalidation (`--force-overwrite`)**: When the text encoder checkpoint or extraction layer changes, the metadata fingerprint mismatches. Explicit `--force-overwrite` is required to wipe and recompute.
4. **Runtime Coverage Check (Fail-Fast)**: If `use_precomputed_text_embeds: true`, the dataset checks 100% prompt coverage at initialization. If even 1 prompt is missing, it raises an actionable error before step 0, rather than crashing mid-training.

#### C. Compute Cost Estimation
* Since feature extraction only requires a single forward pass without autoregressive decoding, throughput on a single H20 / A100 is **~10ms per prompt** (`batch_size=16`).
* **10,000 prompts take ~2 minutes on 1 GPU**, or ~15 seconds using 8 GPUs. Recomputation cost is negligible.

---

## Open Questions & Design Choices

We would love feedback from the maintainers and community on the following architectural choices:

### Q1: Storage Format: Parquet Binary Column vs. Standalone Safetensors?
* **Option A: Embedded Parquet Column (`prompt_embed` as binary/tensor column)**
  * *Pros*: Self-contained within the dataset file; naturally versioned alongside raw prompts and metadata; plays well with Hugging Face datasets and cloud storage.
  * *Cons*: Parquet files grow larger; reading requires row deserialization in Python.
* **Option B: Standalone `.safetensors` sidecar file (Key-Value Tensor Store)**
  * *Pros*: Native zero-copy `mmap` via `safetensors.torch.load_file(..., device="cpu")`; all 8 local ranks share the exact same Linux Page Cache pages without per-process memory duplication.
  * *Cons*: Must be managed and passed as a separate sidecar file.

### Q2: Cluster Distribution & Storage Protocol
* How should precomputed embedding artifacts be shipped to training worker nodes in large clusters?
  1. Stored alongside the dataset on shared cluster storage (e.g. CephFS / HDFS / Lustre / S3 mount)?
  2. Published as a dataset revision artifact on Hugging Face Hub / ModelScope for automatic downloading?
  3. Pre-synced to node-local NVMe during node initialization?

### Q3: Ingestion UX: Dedicated Offline CLI vs. Training-side Auto-Cache?
* Should we provide:
  * **A dedicated offline CLI tool** (e.g. `python -m unirl.tools.precompute_minimax_h3`) where precomputation is strictly decoupled from training?
  * **An automatic run-once-and-save cache mode** during training (e.g., if a cache directory is configured, worker rank 0 computes missing prompts on the first pass and persists them to disk, while subsequent epochs or reruns hit the cache directly)?
  * **A hybrid approach** (CLI for full-scale production training runs, training-side auto-caching for exploratory / interactive jobs)?

### Q4: Extension to Other Heavy Models
* While Minimax-H3 (32B) is the most pressing bottleneck, could this precomputation abstraction be extended to other heavy text encoders in UniRL?
  * **WAN 2.1 / 2.2**: UMT5 text encoder.
  * **SD 3.5 / Flux**: T5-XXL + CLIP.

### Q5: Extension to Multimodal Inputs & Reward Scorers
* Beyond text conditioning:
  * **Reference Image/Video VAE Latents in Editing Tasks**: In image editing tasks (e.g. BAGEL it2i #304) and V2V workflows, encoding the reference image or video frames through the VAE is also a read-only, repeated operation during Rollout. Could reference VAE latents also be precomputed and ingested through a similar mechanism?
  * **Reward Models**: Prompt text embeddings in reward scorers (e.g. VideoPickScore, CLAP) are static and repeatedly evaluated—could they share this precomputation path to reduce evaluation VRAM and latency?

**Comment by zzhuoxin1508** [comment]
Thanks for putting together this RFC — this is well worth doing. 
Two things to watch out for when implementing:
1. The pipeline can't read precomputed embeddings yet. MiniMaxH3Pipeline.generate() computes them unconditionally on the fly:
at pipeline.py L114–120
texts = next((c for c in conditioning if isinstance(c, Texts)), None)
require(texts is not None, "MiniMaxH3Pipeline.generate: no text prompt in the sample conditioning")
conditions = MiniMaxH3Conditions(text=self.text_embed.embed(texts))  # always computed on the fly
Even if the dataset has the embeddings ready, the pipeline won't read them. So we first need to add a bypass — if the conditions already carry an embedding, use it directly; only encode when they don't. This is a prerequisite for precomputation.

2. On whether we need length-based bucketing. H3 currently caps batch size at 1 — forward / micro batch are both 1 — so for now there's no variable-length alignment to worry about; each embedding can just be stored and fetched on its own. If we support batch > 1 later, padding alone won't be enough; we'll also need length bucketing or a repacked layout, which we can handle then.

**Comment by zzhuoxin1508** [comment]
As for the open questions, my leanings:

Q1 (storage) → safetensors,  the sharded layout in NeMo's offline_cache ([ref](https://docs.nvidia.com/nemo/automodel/latest/nemo-automodel/nemo_automodel/components/datasets/llm/offline_cache.html))is a good reference to draw on，don't need to match it exactly.
Q2 (distribution) → shared cluster storage (CephFS), read in place. No moving around. If cold reads get slow, we can add a node-local NVMe pre-sync later; not needed now.
Q3 (ingestion) → a standalone offline CLI is enough. Controllable, and it's the natural place to run the numerical parity check. Training-side auto-cache still loads the 32B encoder on the first epoch, so it doesn't fix the root pain — skip it for now.
Q4 / Q5 (other encoders / VAE latents / reward embeddings) → hold off for now. Let's get the H3 text encoder path working and see how it performs first, then consider extending to the rest.

**Comment by zzhuoxin1508** [comment]
Really nice work here. 
Are you on WeChat by any chance? Would be much easier to discuss than in PR comments.
